### PR TITLE
Revert "Revert "stash: ensure objects and latest version protos have matching hashes

### PR DESCRIPTION
**Pull request by guswynn**
restoring https://github.com/MaterializeInc/materialize/pull/20956, but with a new commit that fixes cargo caching

### CODE REVIEW

File: src/stash/build.rs [hunk 1]
```diff
@@ -226,7 +229,7 @@ fn main() -> anyhow::Result<()> {
     // else has changed.
     //
     // We also rewrite the file if it has been put out of order.
-    if any_new || sorted_to_persist != to_persist {
+    if any_new || dbg!(sorted_to_persist != hashes) {
```
Comment (mjibson) [inline src/stash/build.rs:232]: dbg leftover

File: src/stash/build.rs [hunk 2]
```diff
@@ -101,8 +101,11 @@ fn main() -> anyhow::Result<()> {
     let hashes = fs::File::open(PROTO_HASHES).context("opening proto hashes")?;
     let reader = BufReader::new(&hashes);
     let hashes: Vec<ProtoHash> = serde_json::from_reader(reader)?;
-    let mut persisted: BTreeMap<String, String> =
-        hashes.into_iter().map(|e| (e.name, e.md5)).collect();
+    let mut persisted: BTreeMap<String, String> = hashes
```
Comment (mjibson) [inline src/stash/build.rs:104]: I don't see how this changes anything. Is the actual fix changing the comparison below?

File: src/stash/build.rs [hunk 3]
```diff
@@ -101,49 +101,72 @@ fn main() -> anyhow::Result<()> {
     let hashes = fs::File::open(PROTO_HASHES).context("opening proto hashes")?;
```
Comment (ParkMyCar) [inline src/stash/build.rs:101]: I think we can fix the rebuild issue if we stick
```
println!("cargo:rerun-if-changed=protos");
```
at the beginning of `main`.

According to the [docs](https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorerun-if-changedpath) the default behavior is to re-run the build script if anything in the package changes, and in this case I'm wondering if it's actually anything in our workspace. By adding the `rerun-if-changed` directive it should only re-run if something in the protos directory changes.

Comment (ParkMyCar) [inline src/stash/build.rs:101]: I think we can fix the rebuild issue if we stick
```
println!("cargo:rerun-if-changed=protos");
```
at the beginning of `main`.

According to the [docs](https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorerun-if-changedpath) the default behavior is to re-run the build script if anything in the package changes, and in this case I'm wondering if it's actually anything in our workspace. By adding the `rerun-if-changed` directive it should only re-run if something in the protos directory changes.

File: src/stash/build.rs (hunk 2 above)
Comment (guswynn) [reply src/stash/build.rs:104]: yep!

Comment (guswynn) [reply src/stash/build.rs:104]: I was comparing the wrong object to the sorted object

File: src/stash/build.rs (hunk 3 above)
Comment (guswynn) [reply src/stash/build.rs:101]: do you want to do this in addition to the other fix I made? I think its probably worth it, to avoid extra file io

**Review by guswynn — COMMENTED** [review]
this ended up being not quite worth it, but ill revisit it if there are issues in the future

**Pull request closed without merging**